### PR TITLE
Add RFC for saving accounts to ledger

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Accounts are a form of persistent storage, but that data needs to be stored in memory on each validator. That might be prohibitively expensive if the data needs to stay around for a very long time. The user is effectively renting memory when they'd prefer to buy a little disk space.

#### Summary of Changes

Describes a method for moving account data to the ledger and then restoring it later. The method can be used with no more functionality than what is needed to write a block explorer.

### CODE REVIEW

File: rfcs/0011-longterm-storage.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# Long-term storage
+
+## Background
+
+Accounts act as a form of persistent storage, allowing state to persist across
+multiple transactions. To store data in accounts, however, is relatively
+expensive compared to storing transactions on the ledger. Account storage
+requires memory on *every* validator whereas transaction storage requires only
+disk space on a handful of replicators. If a client does not plan to use
```
Comment (aeyakovenko) [inline rfcs/0011-longterm-storage.md]: Replication taget could be large, like 1000.

Comment (garious) [reply rfcs/0011-longterm-storage.md]: Sure. I suppose I can simply make the argument that disk is cheaper than memory.

**Comment by aeyakovenko** [comment]
From what I can tell this rfc doesn't really cover programs automatically doing this,  or ensuring load/store is verified by the program to be consistent.

File: rfcs/0011-longterm-storage.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+Transaction {
+   keys: vec![
+       my_account,
+       2gVkYWexTHR5Hb2aLeQN3tnngvWzisFKXDUPrgMHpdST,
+   ],
+   instructions: vec![
+       SystemInstruction::Move { tokens: 98 },
+       LoaderInstruction::Write { offset: 0, bytes: [3,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,20,0,0,0,0,0,0,0,50,48,53,48,45,48,49,45,48,49,84,48,48,58,48,48,58,48,48,90,252,10,7,28,246,140,88,177,98,82,10,227,89,81,18,30,194,101,199,16,11,73,133,20,246,62,114,39,20,113,189,32,50,0,0,0,0,0,0,0,247,15,36,102,167,83,225,42,133,127,82,34,36,224,207,130,109,230,224,188,163,33,213,13,5,117,211,251,65,159,197,51,0,0,0,0,0,0]},
+       LoaderIntruction::Finalize,
+       SystemInstruction::Assign { program_id: [1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0] },
```
Comment (aeyakovenko) [inline rfcs/0011-longterm-storage.md]: Assignment can only happen once.  So account `2gVkYWexTHR5Hb2aLeQN3tnngvWzisFKXDUPrgMHpdST` is spent.

File: rfcs/0011-longterm-storage.md (hunk 1 above)
Comment (aeyakovenko) [reply rfcs/0011-longterm-storage.md]: Yea, just a nit.  Handful implies a small amount of data.

File: rfcs/0011-longterm-storage.md [hunk 3]
```diff
@@ -92,14 +93,40 @@ Transaction {
...
+
+1. The API could be reduced to `Store(Pubkey)` and `Load(Pubkey)`
+2. The on-chain program could guarentee the restored version is identical
+   to the original
+3. One less transaction fee
+4. Could be utilized by any program to automatically reduce the cost of using it
+
+### Cannot restore owner
+
+It's possible the proposed solution only work for accounts owned by the
```
Comment (CriesofCarrots) [inline rfcs/0011-longterm-storage.md]: Missing word? "the proposed solution would only work"

Comment (garious) [reply rfcs/0011-longterm-storage.md]: Fixed, thanks!

**Comment by garious** [comment]
@aeyakovenko, care to take another pass?

**Comment by aeyakovenko** [comment]
@garious I am not sure this adds any value.  Seems like all its doing is downloading the account data and deleting the account.  

It needs a protocol for saving the hash of the state on chain and then recovering the data back into the state and asserting that the hash is the same.

**Comment by garious** [comment]
@aeyakovenko, sounds like you missed a critical bit. After downloading the account, the RFC describes a technique for saving that data to the ledger.

**Comment by aeyakovenko** [comment]
@garious But I don't see a way how this would work with existing programs.  The Account userdata's that are owned by the program memory can't change unless it is by the program code only.  The user can't spend the Account's coins to 0 unless the program allows it.

I think this RFC needs to define way for the Program to reduce an Account's userdata to a hash (some fingerprint) of the userdata, and recover the state that matches that hash (fingerprint).  To actually get some savings out of this the OS  would need to provide a way for the Program to set the "used size" of the userdata vector.  Then the OS can "resize" the vector and store fewer bits in memory.

**Comment by garious** [comment]
Great points, I agree! Closing this one for now.

**Pull request closed without merging**